ooni/agent/scheduler.py

```diff
@@ -0,0 +1,409 @@
+import os
+import errno
+
+from hashlib import md5
+from datetime import datetime
+
+from twisted.application import service
+from twisted.internet import defer
+from twisted.internet.task import LoopingCall
+from twisted.python.filepath import FilePath
+
+from ooni.scripts import oonireport
+from ooni import resources
+from ooni.utils import log, SHORT_DATE
+from ooni.utils.files import human_size_to_bytes, directory_usage
+from ooni.deck.store import input_store, deck_store, DEFAULT_DECKS
+from ooni.settings import config
+from ooni.contrib import croniter
+from ooni.contrib.dateutil.tz import tz
+
+from ooni.geoip import probe_ip
+from ooni.measurements import list_measurements
+
+class FileSystemlockAndMutex(object):
+    """
+    This is a lock that is both a mutex lock and also on filesystem.
+    When you acquire it, it will first block on the mutex lock and then
+    once that is released it will attempt to acquire the lock on the
+    filesystem.
+
+    It's a way to support concurrent usage of the DeferredFilesystemLock
+    without races.
+    """
+    def __init__(self, file_path):
+        self._fs_lock = defer.DeferredFilesystemLock(file_path)
+        self._mutex = defer.DeferredLock()
+
+    @defer.inlineCallbacks
+    def acquire(self):
+        yield self._mutex.acquire()
+        yield self._fs_lock.deferUntilLocked()
+
+    def release(self):
+        self._fs_lock.unlock()
+        self._mutex.release()
+
+    @property
+    def locked(self):
+        return self._mutex.locked or self._fs_lock.locked
+
+# We use this date to indicate that the scheduled task has never run.
+# Easter egg, try to see what is special about this date :)?
+CANARY_DATE = datetime(1957, 8, 4, tzinfo=tz.tzutc())
+
+
+class DidNotRun(Exception):
+    pass
+
+
+class ScheduledTask(object):
+    _time_format = "%Y-%m-%dT%H:%M:%SZ"
+    schedule = None
+    identifier = None
+
+    def __init__(self, schedule=None, identifier=None,
+                 scheduler_directory=config.scheduler_directory):
+        if schedule is not None:
+            self.schedule = schedule
+        if identifier is not None:
+            self.identifier = identifier
```

Comment:
Okay, let me see if now I understand what is happening here: in case 1 those two attributes are proper class attributes whereas in case 2 they are instance attributes (meaning that in case 2 each instance gets different values for those attributes), is this correct?

With this in mind, I now understand how the code is written as it is. Now, I have a question: how does class attributes work? When the class is created, they are copied in the `__dict__` underlying a Python instance? Or does `__getattr__` look into the instance `__dict__` and, if that fails, it then looks into the class `__dict__`? I would say the former, because otherwise I do not clearly see how different instances could have different values for these attributes.

To say the same with other words: I find it strange that you can modify "class" attributes for specific instances without affecting other instances. How is that possible? I was thinking that class attributes are like the `prototype` in JavaScript in that the same values of the class attributes are shared by all instances of a class, but apparently it's not like so.

/me still puzzled